[1단계 - 페이먼츠 모듈] 클레어(안은소) 미션 제출 합니다.

eunsoA opened the pull request:
안녕하세요 도밥! 🥸
이번 페이먼츠 모듈 미션의 리뷰이, 클레어입니다! 🍀

이번 미션은 카멜과 함께 페어로 진행했어요 🐪
npm으로 모듈화된 컴포넌트를 만들어보고 npm에 배포하는 경험은 이번 미션으로 처음 해봤는데요, 상당히 재밌었어요. 진짜 개발자가 된 기분이랄까요 ㅎㅎ

그럼 이번 미션 잘 부탁드립니다! 🙇‍♂️


## 🎯 페이먼츠 모듈

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계 - Module

- 재사용 가능한 모듈화된 컴포넌트를 개발하고 npm에 배포할 수 있다.
- 재사용 가능한 커스텀 훅을 개발하고 npm에 배포할 수 있다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 문서화하고 테스트할 수 있다.
- React Testing Library를 사용하여 컴포넌트와 커스텀 훅에 대해 폭넓은 테스트를 작성할 수 있다.

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항(모달 컴포넌트 및 커스텀 훅 개발, npm 배포)을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] Storybook과 RTL 테스트 케이스를 모두 작성했나요?
- [x] 코드 컨벤션을 지키고, 에러 핸들링을 고려했나요?
- [x] 배포한 npm 패키지가 정상적으로 설치 및 작동하나요?
  - 배포 npm 링크: 
    - **[Modal Component 배포 npm](https://www.npmjs.com/package/laireca-modal-components)**
    - **[Custom Hook 배포 npm](https://www.npmjs.com/package/laireca-custom-hooks)**
- [x] Storybook 데모 페이지에 정상적으로 접근할 수 있나요?
  - Storybook 링크: 
    - **[Modal Component Stroybook](https://6813247be99132e63d97379c-ajxxuplmbn.chromatic.com/)**

### 리뷰 요청 & 논의하고 싶은 내용

### 🔽 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

#### 🌀 1. CSS 구현 과정 : CSS Module 대신 Emotion 사용
라이브러리를 배포하기 위해서는 다른 라이브러리와 호환성이 중요하다고 생각했고, 외부 의존성을 최소화 하기 위해 라이브러리를 사용하지 않는 방향으로 CSS Module을 사용할 생각이었어요.

그러나 배포 과정에서 CSS Module을 사용하게 되면, CSS 가 적용되지 않는다는 크루원들의 말을 듣고, 저희는 Emtion을 사용하였어요. 

Emotion과 같은 CSS-in-JS 라이브러리는 스타일을 JavaScript와 함께 번들링하기 때문에, 별도의 CSS 파일 로딩 문제가 발생하지 않는다고 하는데요, 하지만 CSS Module을 사용했을 때, 왜 CSS가 적용되지 않는 문제가 발생하는지에 대한 이유는 명확하게 모르겠어요. 이번 미션 리팩터링을 진행하면서 추가적으로 확인해볼 생각이예요.


#### 🌀 2. 이미지 파일
저는 이전 미션에서 항상 public/ 폴더하위에서 svg 파일 또는 png파일을 관리하였는데요, 이번 미션에서는 페어와 함께 반응형을 고려하여, large / medium / small 중 하나를 Props로 받아서 사용해봤어요. 
반응형을 생각하면 유연한 구조라는 생각이 드는데, 이런 방식으로 처음 만들어봐서 이에 대한 도밥의 생각이 궁금해요!
```tsx
const CloseIconLg = () => (
  <svg></svg>
);

const CloseIconMd = () => (
  <svg></svg>
);

const CloseIconSVG = ({ sizeName = 'md' }: { sizeName?: 'lg' | 'md' }) => {
  return sizeName === 'lg' ? <CloseIconLg /> : <CloseIconMd />;
};

export default CloseIconSVG;
// 자세한 내용은 components/src/lib/assets/CloseIconSVG 에서 확인 가능합니다.
```



### 🔽 2) 이번 리뷰를 통해 논의하고 싶은 부분
#### 🪴 1. (Modal Compnents) 모듈로 배포할 컴포넌트는 어디까지 커스텀 가능해야 할까요?
처음 페어와 미션의 요구사항을 읽고 모듈로 배포할 모달 컴포넌트가 무엇을 받아 어떤 것을 보여 줄지에 대해 많이 고민했어요. 페어는 모달의 배경 색상, 버튼의 색상 등 스타일에 관한 요소까지 커스텀 가능해야 재사용성이 높은 라이브러리 모듈이 될 것 같다고 이야기했고, 저는 반대로 하나의 스타일을 갖고 있어서 간단한 로직만으로도 사용할 수 있는 것이 라이브러리의 하나의 목적이라고 생각했어요.

이런 고민 끝에 현재 저희는 스타일은 고정하고 아래와 같은 구조로 props를 받아서 모달을 그릴 수 있게 하였어요.
| Prop                     | Type                     | Description                                  |
| ------------------------ | ------------------------ | -------------------------------------------- |
| `modalType`              | `'center'` \| `'bottom'` | 모달 위치 설정                               |
| `titleText` _(optional)_ | `string`                 | 모달의 제목                                  |
| `children`               | `ReactNode`              | 모달에 들어갈 내용                           |
| `closeType`              | `'top'` \| `'bottom'`    | 닫기 버튼 위치 설정                          |
| `onClose` _(optional)_   | `() => void`             | 모달이 닫히기 전에 실행되는 사용자 정의 함수 |

위의 입력받은 props를 기반으로 `ModalComponent` 라는 모달 컴포넌트와 해당 모달을 열을 수 있는 `useModal`이라는 커스텀 훅을 반환하고 있어요.

이에 대한 도밥의 생각이 궁금해요!


#### 🪴 2. (Modal Compnents) contextAPI 사용
이번에 modal components의 열고 닫힘을 상태로 관리하기 위해 contextAPI를 사용했어요. 이전에 contextAPI를 사용해본 적이 없어서 페어와 같이 찾아보며 구현을 했었는데요, Provider로 감싸줘야하는 구조이다 보니,` <Modal>` 을 `<ModalProvider>`로 한번더 감싸서 `<ModalComponent>`를 반환하고 있어요. 

라이브러리로 배포해서 사용하는 컴포넌트라면, Modal만 반환하는것이 이상적인 모습이라고 생각했는데요, 리팩터링으로 더 나은 방향으로 수정할 수 있을지 도밥의 생각이 궁금해요!

#### 🪴 3. (custom hooks) payments custom hooks
현재 hooks 모듈에서는 결제 카드 정보 입력에 필요한 다양한 커스텀 훅(`useCardNumbersInput`, `useCardPasswordInput`, `useCardCVCInput`, `useCardExpDateInput`, `useCardCompanyInput`)을 제공하고 있어요.

각 훅은 입력값 관리, 유효성 검증, 에러 메시지 처리 등의 기능을 담당하여 사용자가 카드 정보 입력 폼을 쉽게 구현할 수 있도록 했어요.
이렇게 다양한 훅으로 분리해서 제공하는 것이 좋을지, 아니면 하나의 통합된 useCardForm 같은 훅으로 제공하는 것이 더 나을지 고민이 있었어요. 분리된 훅으로 제공하면 사용자가 필요한 부분만 선택해서 사용할 수 있는 유연성이 있지만, 통합된 훅은 사용성이 더 간편할 수 있을 것 같아요.

또한 훅에서 제공하는 기능의 범위도 고민되었습니다. 현재는 입력값 관리와 유효성 검증 정도만 제공하고 있는데, 더 많은 기능(예: 포맷팅)을 포함해야 할지, 아니면 최소한의 기능만 제공하고 확장성을 높이는 것이 좋을지도 궁금해요.

이런 hooks 모듈의 설계 방향에 대한 도밥의 생각이 궁금해요!

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 모달 컴포넌트의 유연성

- 다양한 위치(중앙, 하단 등)와 내용을 prop으로 받아 모달을 유연하게 구성할 수 있는가?
- 열기, 닫기, 확인 등의 사용자 정의 이벤트 핸들러가 자연스럽게 동작하는가?
- Storybook을 통해 다양한 시나리오를 명확히 테스트했는가?

### 2. 커스텀 훅의 역할과 모듈화

- 카드 정보 입력 유효성 검사 훅이 각각 명확한 책임을 가지도록 분리되어 있는가? 
- 훅이 유효성 검사 결과와 에러 메시지를 명확히 반환하는가?
- 독립적인 테스트 케이스를 작성하여 정상/비정상 입력 모두를 폭넓게 검증했는가?

### 3. 모듈 배포
- npm 배포를 위한 설정이 적절히 구성되어 있는가?

eunsoA committed:
chore: coauthor 설정

Co-authored-by: {dev-dino22} <wlsskdud02@naver.com>

eunsoA committed:
chore: dev 스크립트에 호스트 설정 추가

Co-authored-by: {dev-dino22} <wlsskdud02@naver.com>

eunsoA committed:
fix: coauthor 형식 수정

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
docs: README 기능 명세, 목록 및 작업 내용 작성

Co-authored-by: dev-dino22 <wlsskdud02@naver.com>

eunsoA committed:
chore: @emotion 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
style: App.css에서 불필요한 스타일 제거 및 전체 화면 크기로 변경

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: CloseIconSVG 컴포넌트 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
style: index.css에서 배경색을 흰색으로 변경하고, 전체 화면 크기로 설정

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
style: 모달 스타일 및 모달 텍스트 스타일 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: Modal 및 ModalHeader 컴포넌트 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: App 컴포넌트에 모달 추가 및 레이아웃 구성

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
style: index.css 버튼 스타일 수정 및 hover, focus 상태 변경

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
refactor: CloseIconSVG 컴포넌트 위치 변경

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
style: 모달 스타일 파일 구조 변경 및 ModalTypeProps에 따른 스타일 변경

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: 모달 타입 및 프로퍼티 정의를 위한 modalTypes.ts 파일 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: 모달 열기/닫기 상태 관리 및 컨텍스트 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: 모달 컴포넌트 구조 개선 및 ModalBox, ModalComponent 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: 모달 컴포넌트에서 사용되는 index 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: App 컴포넌트에 버튼 추가 및 모달 프로퍼티 전달

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
test: App 및 Modal 컴포넌트에 대한 스토리북 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
chore: tsconfig 및 vite.config 파일 수정하여 빌드 설정 및 타입 정의 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
refactor: 사용하지 않는 Modal 스토리북 파일 삭제

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
chore: 패키지 이름 및 버전 변경, 설명 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: 카드 입력 유효성 검사 및 오류 메시지 처리 기능 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: 카드 번호 입력 훅 및 테스트 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: 카드 유효기간 입력 훅 및 테스트 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: 카드사 입력 훅 및 테스트 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: 카드 CVC 입력 훅 및 테스트 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: 카드 비밀번호 입력 훅 및 테스트 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: 카드 입력 관련 훅들을 index.ts 파일에 추가

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
feat: 카드 입력 커스텀 훅을 App.tsx에서 사용

Co-authored-by: dev-dino22 <dev-dino22@users.noreply.github.com>

eunsoA committed:
docs: README.md에 모달 컴포넌트 및 사용법 추가

Co-authored-by: dev-dino22 <wlsskdud02@naver.com>

eunsoA committed:
docs: README.md에 카드 결제 관련 커스텀 훅 사용법 및 예시 추가

Co-authored-by: dev-dino22 <wlsskdud02@naver.com>

eunsoA committed:
chore: tsconfig 및 vite.config 수정하여 타입 선언 및 빌드 설정 추가

Co-authored-by: dev-dino22 <wlsskdud02@naver.com>

eunsoA committed:
chore: hooks 패키지 이름 변경 및 설명 업데이트

Co-authored-by: dev-dino22 <wlsskdud02@naver.com>

eunsoA committed:
chore: components 패키지 버전 0.0.1로 업데이트 및 불필요한 publishConfig 제거

Co-authored-by: dev-dino22 <wlsskdud02@naver.com>

eunsoA committed:
chore: 모달 컴포넌트의 임포트 경로를 'laireca-modal-components'로 변경

Co-authored-by: dev-dino22 <wlsskdud02@naver.com>

eunsoA committed:
refactor: InputField 컴포넌트에서 description 프로퍼티 제거

Co-authored-by: dev-dino22 <wlsskdud02@naver.com>

eunsoA committed:
chore(hooks): App.tsx에서 'lib' 임포트 경로를 'laireca-custom-hooks'로 변경

Co-authored-by: dev-dino22 <wlsskdud02@naver.com>

imakerjun reviewed (commented):
@coderabbitai Based on commit history, identify the **recurring work patterns** this developer uses to solve problems.

Focus on these two key questions:

1. **What consistent workflow or cycle does the developer follow?**
    
    – For example: feature → styling → testing. Explain with specific commit examples.
    
2. **What does this pattern suggest about the developer’s mindset or strategy?**
    
    – e.g., feature-first, cleanup-later, test-driven, etc.
    

Structure the result with a **clear title + example + interpretation**, and provide the answer **in Korean**.

wmakerjun commented:
@coderabbitai Based on commit history, identify the **recurring work patterns** this developer uses to solve problems.

Focus on these two key questions:

1. **What consistent workflow or cycle does the developer follow?**
    
    – For example: feature → styling → testing. Explain with specific commit examples.
    
2. **What does this pattern suggest about the developer’s mindset or strategy?**
    
    – e.g., feature-first, cleanup-later, test-driven, etc.
    

Structure the result with a **clear title + example + interpretation**, and provide the answer **in Korean**.

Creative-Lee reviewed (changes requested):
안녕하세요 클레어 👋  반가워요
페이먼츠 모듈 미션 리뷰어 도밥입니다. 잘 부탁드려요😄 

---

간단하게 고민하셨던 부분을 살펴보면,

> #### 🌀 1. CSS 구현 과정 : CSS Module 대신 Emotion 사용

호환성까지 생각해주신 부분을 칭찬드리고 싶네요 👍 
적용되지 않는 이슈에 대해서는 시간이 여유로울 떄 가볍게 한번 파악해보셔도 좋겠습니다 ㅎㅎ
(자칫 깊어질 수 있으니, 뭔가 복잡하게 느껴진다면 바로 빠져나오세요!)

> #### 🌀 2. 이미지 파일 관리

유저향의 prop이 아닌, 개발자가 사용할 prop인것으로 이해했어요. 
의도에 따라 코멘트가 달라질 것 같은데요~!
현재 구현이 step2 까지만 고려해서 개발자가 그외의 사이즈는 사용할 수 없도록 강제하기 위한 의도였다면,(이 라이브러리를 더 확장시켜 나가더라도 x 아이콘은 항상 이 사이즈에서 선택해야해!) 괜찮은 것 같아요.
이런 의도가 아니라면 굳이 제한할 이유가 없기에 svg 관련 props을 전부 받도록 열어 놓는 방식이 더 나을 것 같습니다.
(이 방식에 대해서는 나~중에 개인적으로 학습해보셔도 좋을것 같네요 - 지금은 괜찮아요)

---

질문 주신 부분도 살펴보면,
> #### 🪴 1. (Modal Compnents) 모듈로 배포할 컴포넌트는 어디까지 커스텀 가능해야 할까요?

모듈을 설계할 때 반드시 하게되는 고민이고, 좋은 고민 해주셨어요 👍 
둘의 의견이 뭐하나 옳다고 할 수 없는 부분이라고 생각합니다.
(전자의 경우는 보다 범용적인 사용자에게 제공될 목적에 가깝고, 
후자의 경우는 사내에서 사용하는 디자인이 정해진(내부 요소들도) 경우라고 생각해볼 수도 있겠네요.) 

지금의 모습으로도 children prop을 통해 적당히 사용 가능해보여요.
이어지는 step2에서도 비슷한 고민을 계속하게 될텐데요~!
다양한 고민을 해보고 구현체를 개선해나가면서 느끼는 부분이 많을거에요.
어떻게 하면 커스텀 용이함과 직관적인 사용성을 둘다 챙길 수 있을까? 등의 고민을 해보시는것도 좋겠습니다 (e.g. 이런 요구사항이 생기면 지금 내 모달은 대응이 가능한가? )

>#### 🪴 2. (Modal Compnents) contextAPI 사용
>이번에 modal components의 열고 닫힘을 상태로 관리하기 위해 contextAPI를 사용했어요. 이전에 contextAPI를 사용해본 적이 없어서 페어와 같이 찾아보며 구현을 했었는데요, Provider로 감싸줘야하는 구조이다 보니,` <Modal>` 을 `<ModalProvider>`로 한번더 감싸서 `<ModalComponent>`를 반환하고 있어요. 
라이브러리로 배포해서 사용하는 컴포넌트라면, Modal만 반환하는것이 이상적인 모습이라고 생각했는데요, 리팩터링으로 더 나은 방향으로 수정할 수 있을지 도밥의 생각이 궁금해요!

`열고 닫힘 상태를 외부에서 주입하는게 아닌, 내부적으로 관리하고 싶었다`는 의도였을 것 같은데요. 아직 contextApi의 사용과 개념에 익숙하지 않아서 방법에 어색함이 보이는 것 같아요.
사용법에 대해서는 [요 포스트를 한번 읽고](https://velog.io/@velopert/react-context-tutorial
) 감을 익히시면 좋겠습니다ㅎㅎ 
코드에도 관련된 코멘트 달아두었어요~! 

> #### 🪴 3. payments custom hooks 
>이렇게 다양한 훅으로 분리해서 제공하는 것이 좋을지, 아니면 하나의 통합된 useCardForm 같은 훅으로 제공하는 것이 더 나을지 고민이 있었어요. 분리된 훅으로 제공하면 사용자가 필요한 부분만 선택해서 사용할 수 있는 유연성이 있지만, 통합된 훅은 사용성이 더 간편할 수 있을 것 같아요.

각 훅이 개별로 불리는 케이스가 있을지 없을지는 잘 모르겠지만, 얼핏 생각해보면 또 있을것 같기도 해서 개별 훅은 훅대로 export 하는게 납득되는 것 같고요 ㅎㅎ
통합 혹을 만들고, 이 훅도 사용할 수 있도록 제공하는것도 좋은 생각인 것 같아요.
(isAllValid 같은 파생 상태도 만들 수 있을것 같네요 ㅎㅎ )

> 또한 훅에서 제공하는 기능의 범위도 고민되었습니다. 현재는 입력값 관리와 유효성 검증 정도만 제공하고 있는데, 더 많은 기능(예: 포맷팅)을 포함해야 할지, 아니면 최소한의 기능만 제공하고 확장성을 높이는 것이 좋을지도 궁금해요.

모듈에서 제공해야할 기능의 범위(관심사)가 어디까지인지 잘 정의해보시면 좋을 것 같아요.
예를 들어 포맷팅 같은 경우는 UI의 관심사가 아닐까? 고민해볼수도 있겠습니다.


---

사용처에서 컴포넌트의 prop타입이 자동 완성되도록 챙겨주신 부분 좋네요 ㅎㅎ
고생하셨습니다 클레어~! 코멘트 한번 차근히 반영해보셔요 😄

Creative-Lee commented on components/src/lib/assets/CloseIconSVG.tsx:
```diff
@@ -0,0 +1,39 @@
+  </svg>
+);
+
+const CloseIconSVG = ({ sizeName = 'md' }: { sizeName?: 'lg' | 'md' }) => {
+  return sizeName === 'lg' ? <CloseIconLg /> : <CloseIconMd />;
+};
```
중요한 것을 누락해서 뒤늦게 남겨보아요.
위 처럼 svg를 컴포넌트화 해서 사용하려면 `vite-plugin-svgr` 같은 플러그인 셋업이 필요해요!
(코멘트의 의도만 이해했다면, 넘어가셔도 좋습니다~!)

eunsoA replied:
오 사실 아이콘 파일을 어떻게 관리할지 페어와 고민하다가 말씀해주신 대로 컴포넌트로 만들어서 관리하는 방식도 고려했었어요.
그런데 이렇게 svg를 컴포넌트화하여 구현하면 말씀하신대로 `vite-plugin-svgr` 같은 플러그인에 의존성이 추가로 생기더라구요.

페어와 이 문제에 대해 논의한 결과, 저희는 라이브러리 모듈을 만드는 것이 목표인 만큼 다른 의존성이 생기는 것보다는 현재 방식이 더 나을 것 같다는 생각에 이 방식을 채택 했었어요!

만약 라이브러리 모듈로 사용하지 않을 목적이었다면 말씀해주신 대로 svg를 컴포넌트화 하여 사용했을 것 같아요.

도밥이 생각하시기에 이번 미션에서는 svg를 컴포넌트화하여 변경하는게 나을까요? 만약 그렇다면 step2를 진행하면서 반영하겠습니닷!

Creative-Lee replied:
이미 고려를 했던 내용이군요 ㅎㅎ 👍 
학습 겸 svgr플러그인 적용을 해보는것도 좋다고 생각합니다.
다만 미션 전체로 보았을 때는 아주 작은 일부분이라, step2에서 미션의 목표를 모두 달성한 뒤에도 시간이 남는다면 해보는 걸로 하시죠~!

eunsoA committed:
chore: package에 'laireca-modal-components' 제거 및 'chromatic' 추가, react와 react-dom 를 peerDependencies로 변경

eunsoA committed:
chore(hooks): package에서 react와 react-dom을 peerDependencies로 이동

eunsoA committed:
refactor(component): ModalHeader 컴포넌트에서 onClose 옵셔널 체이닝 사용

eunsoA committed:
chore(component): ModalComponent에서 스타일 관련 import 제거

eunsoA committed:
refactor(component): 모달 컨텍스트 관리 구조 개선

- 전역 변수 대신 Context를 통해 모달 상태 관리
- 불필요한 ModalComponent 래퍼 제거, Modal 직접 사용
- useModal 훅 내부에 Context 사용 검증 로직 추가

eunsoA committed:
refactor(component): Context API 활용한 모달 컴포넌트  Props Drilling 문제 해결

- ModalContent, ModalBox, ModalHeader 컴포넌트에서 props를 간소화하여 불필요한 전달 제거
- ModalProvider에서 modalType, titleText, closeType을 직접 관리하도록 변경
- Modal 컴포넌트에서 useModal 훅을 통해 상태 관리 개선

eunsoA committed:
refactor(hooks): 훅 파일명과 함수명 통일

eunsoA committed:
refactor(hooks): 카드 만료일 입력 훅에서 불필요한 errorResult 객체 제거 및 로직 개선

eunsoA committed:
refactor(hooks): getFirstErrorMessage 에서 공통된 내용을 getFirstErrorMessage로 통합

eunsoA committed:
refactor(hooks): useCardCompanyInput 훅에서 errorMessage의 setter 추가

eunsoA committed:
refactor(hooks): 카드 입력 검증 로직에 입력값 길이 검증 추가 및 에러 메시지 업데이트

eunsoA committed:
refactor(hooks): 카드 입력 검증 로직에서 VALIDATION_LENGTH 사용 및 에러 메시지 동적 업데이트

eunsoA committed:
refactor(hooks): 카드 번호 입력 검증 로직에서 카드 번호 전체 길이 검증 추가

eunsoA committed:
refactor(hooks): 카드 입력 관련 훅에서 에러 상태 관리 방식 개선

- 유효성 검사 결과와 에러 메시지를 하나의 객체로 통합하여 관리

eunsoA committed:
chore(component): 버전 업데이트 - package.json 및 package-lock.json에서 버전을 0.0.1에서 0.1.0으로 변경

eunsoA committed:
chore(hooks): 버전 업데이트 - package.json 및 package-lock.json에서 버전을 0.0.1에서 0.1.0으로 변경

eunsoA commented:
안녕하세요 도밥! 연휴 잘 보내셨나요 ⛱️
꼼꼼하게 리뷰해주시고 생각할 내용을 많이 던져주신 덕분에 많이 배웠던 것 같아요! 🙇‍♂️

수정 및 반영한 내용은 크게 아래와 같습니다.

🌀 component
1. 모달 컨텍스트 관리 구조 개선
  - 전역 변수 대신 Context를 통해 모달 상태 관리하도록 수정했어요.

🌀 hooks
1. getFirstErrorMessage 과 getExpirationFirstErrorMessage에서 공통된 내용을 getFirstErrorMessage로 통일
2. 카드 입력 검증 로직에 입력값 길이 검증 추가 
  - e.g. cvc 입력 길이
3. 카드 입력 관련 훅에서 에러 상태 관리 방식 개선
  - errorMessage가 아니라 isVaild와 errorMessage를 갖는 error 객체로 관리하도록 수정했어요.

🌀 라이브러리 관련
1. 버저닝 관련 아티클 내용을 반영하여 수정
(commit : f23863a06ba00ecf21617d97a2f0d834308b4662, 21caa1075df86373c560465ae9416b7eadcb6b44)

또한 위에 코멘트 주신 내용 모두 반영했어요. 추가로 몇가지 궁금한 질문 내용들을 위 코멘트에 남겨놨습니다! (한두개정도예요)
감사합니다 좋은하루 보내세요!🍀

eunsoA commented on components/package.json:
```diff
@@ -1,7 +1,7 @@
 {
-  "name": "compoents",
+  "name": "laireca-modal-components",
```
오 좋은 아티클 감사해요!
해당 글을 읽고 간단하게 정리해봤어요 :]

[시맨틱 버저닝 이해]
- 시맨틱 버저닝 
  : 소프트웨어의 변경 사항을 명확하게 전달하고 호환성을 유지하는 데 도움을 주는 버전 관리 방법
  - 버전은 세 개의 숫자 (메이저.마이너.패치)로 구성됨.
- 각 숫자별 의미:
  - 메이저 버전
    : 큰 기능 추가나 변경을 의미하며, 이전 버전과 호환되지 않는 경우가 많음. 고로 변경 시 신중한 계획과 테스트가 필요함.
  - 마이너 버전
    : 작은 기능 추가나 버그 수정을 의미하며, 일반적으로 이전 마이너 버전과 호환성 문제가 거의 발생하지 않음. 소프트웨어의 안정성을 높이는 데 기여하며 자주 이루어질 수 있음.
  - 패치 버전
    : 사소한 수정이나 주석 변경을 의미하며, 소프트웨어 기능에 큰 영향을 미치지 않고 호환성 문제가 거의 없음. 
    소프트웨어의 품질을 높이는 데 기여하며 자주 이루어질 수 있음.

eunsoA replied:
그리고 추가적으로 안 사실인데 `패키지의 초기 버전을 설정할 때, 0.1.0 을 사용하는 것이 일반적인 관례`라고 하더라구요!

- Semantic Versioning (SemVer) 명세에 따르면, 메이저 버전 0 (예: 0.x.y)은 초기 개발 단계를 의미하며, 이 단계에서는 API가 안정적이지 않고 언제든지 변경될 수 있음을 나타냅니다.
  - 0.1.0: 첫 번째 개발 릴리스를 나타내는 가장 흔한 시작점입니다. "아직 프로덕션 용도는 아니지만, 첫 번째 기능 세트/API 초안이 준비되었다"는 의미를 내포합니다.
  - 0.0.0: 기술적으로 가능하지만, "아직 아무것도 준비되지 않았다" 또는 "버전 관리를 시작하기 전"이라는 의미로 해석될 수 있어 실제 초기 릴리스 버전으로는 잘 사용되지 않습니다.

따라서 새로운 패키지 개발을 시작할 때는 0.1.0으로 시작하여 기능 추가나 변경 시 마이너(0.x.0) 또는 패치(0.1.x) 버전을 올리고, 공개적으로 안정된 API를 처음 릴리스할 준비가 되었을 때 1.0.0으로 올리는 것이 좋습니다.

eunsoA replied:
반영하여 업데이트 했습니다! (commit : f23863a06ba00ecf21617d97a2f0d834308b4662, 21caa1075df86373c560465ae9416b7eadcb6b44)

eunsoA commented on components/package.json:
```diff
@@ -10,14 +10,16 @@
   },
   "dependencies": {
+    "@emotion/css": "^11.13.5",
+    "@emotion/styled": "^11.14.0",
```
네! 아래에 `peerDependencies`와 `dependencies`를 비교해봤어요.

- `dependencies`
  - 프로젝트가 실행되기 위해 직접적으로 필요한 패키지들을 명시함.
  - npm install 실행 시 자동으로 설치됨.
  - 프로젝트의 실행에 필수적인 라이브러리들이 여기에 포함됨.
  - ex. 현재 프로젝트의 @emotion/css, @emotion/styled 같은 스타일링 라이브러리들이 여기에 해당됨.

- `peerDependencies`
  - 프로젝트가 호환되는 다른 패키지의 버전을 명시함.
  - 직접 설치되지 않고, 사용하는 프로젝트에서 해당 패키지를 제공해야 함.
  - 주로 플러그인이나 라이브러리를 만들 때 사용됨.
  - 예를 들어 React 컴포넌트 라이브러리를 만든다면 react와 react-dom을 peerDependencies로 지정하는 것이 좋음.

- 사용해야 하는 경우
  - `dependencies`: 
    - 프로젝트가 직접적으로 사용하는 패키지들 (예: 유틸리티 라이브러리)
  - `peerDependencies`:
    - 여러 버전이 동시에 설치되면 충돌이 발생할 수 있는 패키지들
    - 라이브러리의 인터페이스에 노출되는 패키지들
    - 사용자가 버전을 선택해야 하는 패키지들
 
결론적으로 만약 dependencies로 지정하면, 라이브러리를 사용하는 프로젝트에서 다른 버전의 React를 사용할 때 문제가 발생할 수 있기에, npm 배포시에는 react와 react-dom에 대해서 `peerDependencies`에 명시하는게 좋겠네요!

수정하여 반영했습니다! ( commit :  7bdca21cb58f64c55b4a864750bc7342829d9a03 , d2d856606b2d392ccb0fbe3d400bb4c66f619415)

eunsoA commented on components/package.json:
```diff
@@ -1,7 +1,7 @@
+  "name": "laireca-modal-components",
   "version": "0.0.1",
-  "description": "npm modules",
+  "description": "Easily create modal components with this lightweight npm package",
```
넵 확인해보겠습니다!

eunsoA commented on components/src/lib/index.ts:
앗 기존에 index.ts에서 사용자에게 필요한 `ModalComponent`, `ModalProvider`, `useModal`만 export 하고 있는데요, 그럼에도 외부에서 import 가능한가요?? 

우선 사용되지 않는 style 관련 import 문은 지웠습니다! (commit : 0c98f073e684d77002f22d47f11bf39226a2ab9c)

Creative-Lee replied:
이 부분은 클레어가 직접 빈 프로젝트를 하나 생성해서 클레어의 라이브러리를 설치해보고
import 자동완성이 어떻게 되는지 확인해보는걸로 하시죠~!

eunsoA commented on components/src/lib/contexts/ModalContext.tsx:
```diff
@@ -0,0 +1,19 @@
+  const closeModalHandler = () => _setIsModalOpened(false);
+
+  return { openModalHandler, closeModalHandler };
+};
```
PR 코멘트 본문에 올려주신 contextAPI 관련 아티클을 보고 말씀해주신 내용을 읽어보니, 현재 제 코드의 문제점을 명확하게 짚어주셨네요 ..! ㅎㅎ

페어와 논의하여 context를 사용해보고싶다는 생각에 구현하다보니 아무래도 말씀하신대로 context와 hook의 목적을 잘 이해하지 못하고 사용했던 것 같아요.

단순히 사용자는 `ModalComponent`를 주고 해당 모달을 open & close만 할 수 있게 만들고자 의도했어요. 그러다보니 Provider도 외부에서 사용자가 직접 사용하기보다는 감싸진 형태의 `ModalComponent`의 형태로 반환하고자 했어요.

그런데 말씀해주신 코멘트를 보니 이렇게 되면 모달을 여러개를 띄울때 문제가 생기겠네요 .. 😥

--- 
말씀해주신 내용 반영해서 수정했습니다! 주 변경사안은 아래와 같아요. (commit : b55e6ee4b1ad55b0134114a817f9a07897cf1062 )

1. 전역 변수를 사용하는 방식을 제거하고, Context를 통해 모달 상태와 관련 함수를 제공하도록 변경
    - Context에 isModalOpened, openModalHandler, closeModalHandler를 포함시켜 Modal 상태 관리에 필요한 모든 것을 제공
    - useModal 훅에서 Context를 사용할 때 Provider 내부에서만 사용 가능하도록 에러 처리 추가

2. ModalComponent를 제거하고 Modal을 직접 사용하도록 변경
    - 기존에 ModalComponent 내부에 있던 Provider를 App 컴포넌트로 이동시켜 사용자가 직접 Provider를 사용하도록 변경

3. App 구조를 개선
    - App이 ModalProvider를 감싸고, 그 안에 ModalContent 컴포넌트를 배치하는 형태로 변경

eunsoA replied:
추가적으로 Props Drilling 문제를 Context를 활용하여 개선해봤어요. (commit : 295c1636fa7dc057d58f559ea2987f67b1a42e5d )
  - children만 직접 전달하고 나머지는 useModal 훅을 통해 필요한 값들을 Context에서 가져옴. -> Props Drilling 문제를 발생 방지
  - 모달 관련 상태와 속성(modalType, titleText, closeType, onClose 등)을 모두 Context에서 관리합니다.
  - 컴포넌트 간 props 전달을 최소화하여, 중간 컴포넌트들이 props를 전달만 하는 역할에서 벗어남.

Creative-Lee replied:
https://github.com/woowacourse/react-modules/commit/b55e6ee4b1ad55b0134114a817f9a07897cf1062 
적절하게 잘 반영해주셨습니다 ㅎㅎ 의도가 명확해졌네요

---
https://github.com/woowacourse/react-modules/commit/295c1636fa7dc057d58f559ea2987f67b1a42e5d
이 변경사항에 대해서는 사용자 입장에서도 생각해볼필요가 있어요.
개발자(클레어)는 Props Drilling 문제가 해결되어 로직이 깔끔해 보일 순 있습니다.

하지만 사용자(다른 개발자) 입장에서 바라보면...
- Provider에 modal 관련 옵션을 전달했고,
- 실제 모달의 내용물(children)을 작성하는 `Modal`컴포넌트 사용처는 Provider 내부 어딘가 일테니,

두 컴포넌트는 물리적으로 떨어져 있을 겁니다.
수정 전 구현에서 처럼 `Modal` 컴포넌트에 옵션을 전달할 때 보다 **전체적인 모양새를 한눈에 파악하기 힘들겠죠.** 

더불어 `Modal`과 관련된 컴포넌트들 끼리 2~3depth 정도로 prop을 전달하는 것은 
전형적인 Props Drilling 문제라고 하기에는 다소 거리가 있습니다.

사용자를 위한 인터페이스에 대해서도 생각해보셔요~!

eunsoA commented on components/src/lib/ModalHeader.tsx:
```diff
@@ -0,0 +1,25 @@
+const ModalHeader = ({ titleText, hasCloseButton, onClose }: ModalHeaderProps) => {
+  const { closeModalHandler } = useModal();
+  function closeHandler() {
+    if (onClose) onClose();
```
앗 그렇네요! 감사합니다!
수정해서 반영했습니다! (commit : cc64fbae2a4e260e236ff80599ffc99f02a9d5dc)

eunsoA commented on components/src/lib/Modal.tsx:
```diff
@@ -0,0 +1,49 @@
+import { ModalProps } from './types/modalTypes';
+
+const Modal = ({ modalType, titleText = '', children, closeType, onClose }: ModalProps) => {
+  const context = useContext(ModalContext);
```
넵! 이부분도 위의 답변에서 반영했습니다! (commit : b55e6ee4b1ad55b0134114a817f9a07897cf1062)

eunsoA commented on components/src/lib/types/modalTypes.ts:
```diff
@@ -0,0 +1,19 @@
+  children: ReactNode;
+  closeType: 'top' | 'bottom';
+  onClose?: () => void;
+}
```
현재 모달 컴포넌트는 ModalContext를 통해 내부적으로 상태(isModalOpened)를 관리하고 있는데요! 
관심사 분리의 목적이 컸어요. 모달의 열림/닫힘 상태와 같은 UI 로직을 모달 컴포넌트 내부에 캡슐화하여, 사용하는 쪽에서는 단순히 모달 내용에만 집중하고자 했어요.

외부에서 상태를 주입하는 방식이라는게 명확히 이해가 안되는데 조금 더 자세하게 설명해주실 수 있을까요?? 👀

Creative-Lee replied:
목적이 분명했다면 어떤 구현이던 좋습니다 ~! 👍 

---

`const [isOpen, setIsOpen] = useState(false)` 등의 상태를 
`Modal` 컴포넌트에게 prop으로 전달하여 사용하는 방식의 인터페이스를 말씀드린겁니다! ㅎㅎ
(Modal 내부에서는 `isOpen && <BlahBlah />` 같이 조건부 랜더링을 통해 visible 여부를 제어하는등의 구현)

eunsoA commented on components/src/lib/contexts/ModalContext.tsx:
```diff
@@ -0,0 +1,19 @@
+
+var _setIsModalOpened: Dispatch<SetStateAction<boolean>>;
+
+export const ModalProvider = ({ children }: any) => {
```
옙! 이것도 위의 코멘트를 반영하면서 수정했어요! (commit: b55e6ee4b1ad55b0134114a817f9a07897cf1062 )

eunsoA commented on hooks/src/lib/useCardCVC.ts:
```diff
@@ -0,0 +1,21 @@
+import { validateCVC } from './validator/validateCardInput';
+import { getFirstErrorMessage } from './validator/getFirstErrorMessage';
+
+export function useCardCVCInput() {
```
앗 넵 수정해서 반영했습니다! (commit : ad35d3ecac89fcd5ae5c3ca878d5c372c1e573c1 )

eunsoA commented on hooks/src/lib/useCardCVC.ts:
```diff
@@ -0,0 +1,21 @@
+
+export function useCardCVCInput() {
+  const [cardCVC, setCardCVC] = useState('');
+  const [errorMessage, setErrorMessage] = useState('');
```
넵! 이부분도 수정하여 반영하였습니다! (commit : 194ced59467ea740e9008e9b4423b0ef7834f228)

eunsoA replied:
엇이거 답변이 지워졌네요..? 다시 달자면 ..

넵! 이부분도 수정하여 반영하였습니다! (commit : https://github.com/woowacourse/react-modules/commit/194ced59467ea740e9008e9b4423b0ef7834f228)

eunsoA commented on hooks/src/lib/useCardExpDate.ts:
```diff
@@ -0,0 +1,41 @@
+  const errorResults = {
+    month: {},
+    year: {},
+  };
```
오 그러네요..! 수정해서 반영했습니다! 감사해요 (commit : 11b8668b1611e461bb3f42b33b49f243adb6a2e3 )

eunsoA commented on hooks/src/lib/useCardExpDate.ts:
```diff
@@ -0,0 +1,41 @@
+    setCardExpDate(nextExpDate);
+
+    errorResults.month = validateExpirationDateMonth(nextExpDate.month, nextExpDate.year);
+    errorResults.year = validateExpirationDateYear(nextExpDate.month, nextExpDate.year);
```
위 수정사안과 겹치는 내용이라 동일하게 반영하여 수정했어요! (commit : 11b8668b1611e461bb3f42b33b49f243adb6a2e3)

eunsoA replied:
이것도 답변이 지워져서 다시 달았어요!

위 수정사안과 겹치는 내용이라 동일하게 반영하여 수정했어요! (commit : https://github.com/woowacourse/react-modules/commit/11b8668b1611e461bb3f42b33b49f243adb6a2e3)

eunsoA commented on hooks/src/lib/validator/validateCardInput.ts:
앗 자릿수 유효성 검사에 대해서는 페어프로그래밍 당시 놓쳤던 것 같아요 ..!
추가하여서 반영하였습니다! (commit : 326e41c0f126dc271facacc3fdc3b8f2e957d71f, 73baa4e51e097813a4f649cfae4e5afd47c15bbe, f471a44908c5ab4739d64cb77982f4b94f445fff)

eunsoA commented on hooks/src/lib/validator/getFirstErrorMessage.ts:
앗 좋습니다!
수정해서 반영했습니다! (commit : 9a4667c51f1e81a875a021e60fe2935b81cbde48 )

eunsoA commented on hooks/src/lib/useCardCompany.ts:
```diff
@@ -0,0 +1,17 @@
+    onChangeHandler,
+    errorMessage,
+  };
+}
```
네 CardCompany는 단순 selecter로 에러메세지가 출력될 일이 없어서 setter를 빈값으로 뒀습니다!
혹시 필요할까요??

---
흠 범용성 좋은 모듈이라면 setter가 있어도 좋겠다는 생각은 들어요.
```ts
function onChangeHandler(e: React.ChangeEvent<HTMLSelectElement>) {
    const { value } = e.target;
    setCardCompany(value);
    const errorResult = validateCardCompany(value);
    setErrorMessage(getFirstErrorMessage(errorResult, 'COMPANY'));
  }
```
만약 에러 메세지 setter가 있다면 위와 같겠지만 카드사 정보의 경우 따로 validate로직이 없어서 아래와 같이 수정했습니다! (commit : bfa88625ef1b0c744e6547e8e64810d4ecfb048f )
```ts
  function onChangeHandler(e: React.ChangeEvent<HTMLSelectElement>) {
    const { value } = e.target;
    setCardCompany(value);
    setErrorMessage('');
  }
```

Creative-Lee replied:
step2에서는 필요하게 될지도 모르겠네요 ㅎㅎ

---

별개로 핸들러가 인자로 이벤트 객체 전부를 받아야하는지도 한번 고민해보세요~!
(사용처에서는 'value'만 전달하도록 해도 되지 않을까?)

eunsoA commented on hooks/src/lib/useCardExpDate.ts:
```diff
@@ -0,0 +1,41 @@
+  };
+
+  function onChangeHandler(e: React.ChangeEvent<HTMLInputElement>) {
+    const { name, value } = e.target;
```
아뇨 .. 확신할 수 없어요 ...  

`좋은 라이브러리나 모듈은 쓰는 사람이 실수하기 어렵게 만들어서 의도대로 만도록 해야한다` 라는 말을 봤어요. 사용자 입장에서 실수하기 어렵게 만들려면 작은 단위에서 책임이 분명하고 실수를 덜하게 끔 의도해야할 것 같아요.

여기까지는 알겠는데 그럼 어떻게 해야하지?가 가장 큰 고민이었어요.

위의 모달 컴포넌트에서 context가 선언되지 않은 상태로 useModal을 사용하면 에러를 던지듯이 그러한 트리거?같은 걸 만들면 좋을까요?

Creative-Lee replied:
그런 트리거도 방법이겠고, 간단하게 type을 제한한 `name`을 인자로 받는것도 방법이겠습니다.
```ts

// 사용처의 모습 예시
onChange={(e) => onChangeHandler('month', e.target.value)}
```

eunsoA commented on hooks/src/lib/useCardNumbers.ts:
```diff
@@ -0,0 +1,25 @@
+
+  function onChangeHandler(e: ChangeEvent<HTMLInputElement>) {
+    const { name, value } = e.target;
+    const index = Number(name);
```
이것 또한 위와 이어지는 질문인 것 같아요!

Creative-Lee reviewed (approved):
안녕하세요 클레어 😄 
전체적으로 꽤나 큰 변경사항이 발생할 요청들을 많이 드렸는데, 
학습도 잘 해주셨고, 코멘트도 적절하게 잘 반영해주셔서 놀랐네요ㅎㅎ
마무리 코멘트 잘 확인해주시고, 호다닥 step2 도 진행해보시죠 🏃 🏃 🏃 

화이팅입니다 💪 
고생하셨어요 클레어 👍

eunsoA commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one place** where the code is written in an imperative style, and explain how to make it **more declarative**.

# How to respond  
Start your response with the following markdown title:  
`# 1. ✨ 선언형 UI로서 개선할 피드백 포인트`

Then respond with:  
- **Imperative pattern name or symptom**
- **Where it appears** (files + line ranges or clear examples)  
- **Why it’s not declarative**  
- **AS-IS**: _Show the current (imperative) code snippet_  
- **TO-BE**: _Show the rewritten (declarative) version_  
- **Why this change improves React code quality** (≤ 2 sentences)

# Style  
Respond in bullet points, and include only one clear improvement point.
Please make sure your feedback is based on my code and pull request, and summarize the key improvement point in one or two sentences at the end.

# Language  
Please write your response **in Korean**.

eunsoA commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify exactly one state value that can be derived from other state or props, and explain how removing it would simplify the code.

# How to respond
Start your response with the following markdown title:
`# 2. 📉 상태 최소화를 위한 개선 포인트`

Then respond with:

- Derived state name
- Where it appears (files + line ranges or variable name)
- Why it’s redundant
- AS-IS: Show the redundant state and its logic
- TO-BE: Show how to derive it from existing values
- Why this is the highest-leverage fix (≤ 2 sentences)

# Style
Respond in bullet points, and focus on one state-related improvement.

# Language
Please write your response in Korean.

eunsoA commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one component** whose responsibility is unclear or mixed, and explain how to refactor it to **separate concerns more clearly**.

# How to respond  
Start your response with the following markdown title:  
`# 3. 🧩 컴포넌트 책임 분리를 위한 리팩토링 포인트`

Then respond with:  
- **Component name or file**  
- **Where the responsibility is unclear**  
- **Why this hurts reusability and maintainability**  
- **AS-IS**: _Show the current (overloaded) component logic_  
- **TO-BE**: _Show a refactored structure or extraction example_  
- **Why this is the most impactful structural change** (≤ 2 sentences)

# Style  
Respond in bullet points, and provide clear before/after contrast.

# Language  
Please write your response **in Korean**.

eunsoA commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one recurring mistake, anti-pattern, or code smell** that appears in multiple places and seems to reflect the author's coding habit.  
Explain what it is, why it may keep happening, and how the author can recognize and change this pattern.

# How to respond  
Start your response with the following markdown title:  
`# 4. 🔁 반복적으로 나타나는 안티패턴 또는 코드 습관 피드백`

Then respond with:
- **Pattern or smell name**  
- **Where it appears repeatedly** (files + line ranges or variable/component patterns)  
- **Why it likely keeps happening** (assumed thought process or misconception)  
- **AS-IS**: _Example of current usage or structure_  
- **TO-BE**: _Refactored or improved structure with reasoning_  
- **How the author can self-check next time** (one reflection or checklist question)

# Style  
Respond in bullet points, focused on identifying **a single recurring behavioral pattern**.

# Language  
Please write your response **in Korean**.
